swig/ChunkedArray.hpp

```diff
@@ -0,0 +1,240 @@
+#ifndef __CHUNKED_ARRAY_H__
+#define __CHUNKED_ARRAY_H__
+
+#include <new>
+#include <vector>
+#include <algorithm>
+
+#include <stdint.h>
+#include <assert.h>
+
+/**
+ * Container that manages a dynamic array of fixed-length chunks.
+ *
+ * The class also takes care of allocation & release of the underlying
+ * memory. It can be used with either a high or low-level API.
+ *
+ * The high-level API allocates chunks as needed, manages addresses automatically and keeps
+ * track of number of inserted elements, but is not thread-safe (ok as usually input is a streaming iterator).
+ * For parallel input sources the low-level API must be used.
+ *
+ * Note: When using this for `LGBM_DatasetCreateFromMats` use a
+ *       chunk_size multiple of #num_cols for your dataset, so each chunk
+ *       contains "complete" instances.
+ *
+ * === High-level insert API intro ===
+ *
+ * The easiest way to use is:
+ *  0. ChunkedArray(chunk_size)  # Choose appropriate size
+ *  1. add(value)                # as many times as you want (will generate chunks as needed)
+ *  2. data() or void_data()     # retrieves a T** or void** pointer (useful for `LGBM_DatasetCreateFromMats`).
+ *
+ * Useful query methods (all O(1)):
+ *  - get_add_count()   # total count of added elements
+ *  - get_chunks_count()  # how many chunks are currently allocated.
+ *  - get_current_chunk_added_count()  # for the last add() chunk, how many items there are.
+ *  - get_chunk_size()    # Get constant chunk_size from constructor call.
+ *
+ *  With those you can generate int32_t sizes[]. Last chunk can be smaller than chunk_size, so, for any i:
+ *    - sizes[i<last] = get_chunk_size()
+ *    - sizes[i==last] = get_add_count()
+ *
+ *
+ * === Low-level insert API intro ===
+ *
+ * For advanced usage - useful for inserting in parallel - one can also:
+ *  1. call new_chunk() at any time for as many chunks as needed.  (thread-UNsafe)
+ *  2. call setitem(chunk, idx, value) to insert each value.       (thread-safe)
+ *
+ */
+template <class T>
+class ChunkedArray
+{
+  public:
+    ChunkedArray(size_t chunk_size)
+      : _chunk_size(chunk_size), _current_chunks_idx(0), _current_chunk_idx(0)
+    {
+       new_chunk();
+    }
+
+    ~ChunkedArray()
+    {
+        release();
+    }
+
+    /**
+     * Adds a value to the chunks sequentially.
+     * If the last chunk is full it creates a new one and appends to it.
+     */
+    void add(T value) {
+        if (! within_bounds(_current_chunks_idx, _current_chunk_idx)) {
```

Comment:
nit: space between ! within_bounds
actually, I'm surprised there is no linter that catches such issues

Reply:
I see a similar thing on line 223, surprised there's no linter catching this

Reply:
@imatiach-msft 
> actually, I'm surprised there is no linter that catches such issues

Sorry, we do not run linting tests for `swig` folder.
Refer to 
https://github.com/microsoft/LightGBM/blob/0ee4d37fb59e08bccc60e051764e5996245a0cc5/.ci/test.sh#L67

I remember I tried to do this but `cpplint` produced some number of (possibly false positive) errors and I abandoned that idea. We will really appreciate if you enable cpplint for `swig` folder and fix or mark as `// NOLINT` all errors. If you don't want to install `cpplint` locally you can simply propose a PR with added `swig` folder and check our CI logs for errors.
```
cpplint --filter=-build/c++11,-build/include_subdir,-build/header_guard,-whitespace/line_length --recursive ./src ./include ./R-package ./swig || exit -1
```

Reply:
Already added swig to the cpplint command and fixed all swig errors in the swig folder.

Reply:
> Already added swig to the cpplint command and fixed all swig errors in the swig folder.

You are awesome!